_posts/2025-12-26-Deploying_OCI_Images_with_Custom_IPA_Hardware_Manager.md

```diff
@@ -0,0 +1,1966 @@
+    set -x
+fi
+
+set -eu
+set -o pipefail
+
+# Enable CRB (CodeReady Builder) repository and install EPEL
+echo "Enabling CRB repository..."
+dnf config-manager --set-enabled crb || true
+
+# Detect CentOS version and install appropriate EPEL
+if [ -f /etc/os-release ]; then
+    # shellcheck disable=SC1091
+    . /etc/os-release
+    case "${VERSION_ID%%.*}" in
+        9)
+            echo "Installing EPEL for CentOS 9..."
+            dnf install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-9.noarch.rpm || true
+            ;;
+        10)
+            echo "Installing EPEL for CentOS 10..."
+            dnf install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-10.noarch.rpm || true
+            ;;
+        *)
+            echo "Unknown CentOS version: ${VERSION_ID}, skipping EPEL installation"
+            ;;
+    esac
+fi
+
+if [ -z "${DIB_EXTRA_PACKAGES:-}" ]; then
+    echo "No extra packages specified via DIB_EXTRA_PACKAGES, skipping"
+    exit 0
+fi
+
+echo "Updating system packages..."
+dnf update -y
+
+echo "Installing extra packages: ${DIB_EXTRA_PACKAGES}"
+
+# shellcheck disable=SC2086
+dnf install -y ${DIB_EXTRA_PACKAGES}
+
+echo "Cleaning package cache..."
+dnf clean all
+
+echo "Extra packages installation complete"
+```
+
+  </div>
+</details>
+
+### Building the Image
+
+Set the `ELEMENTS_PATH` to include your custom elements directory, then
+run the builder:
+
+```bash
+export ELEMENTS_PATH="/path/to/your/dib-elements"
+
+export DIB_EXTRA_PACKAGES="jq yq mdadm lvm2 curl parted util-linux \
+    squashfs-tools xfsprogs dosfstools grub2-efi-x64 grub2-tools rsync"
+
+ironic-python-agent-builder \
+    -o ipa-custom \
+    -e extra-hardware \
+    -e crane \
+    -e packages-install \
+    --release 9-stream centos
+```
+
+This produces two files:
+
+- `ipa-custom.kernel` - The Linux kernel
+- `ipa-custom.initramfs` - The ramdisk containing IPA and tools
+
+For ARM64 builds, the grub packages differ:
+
+```bash
+export DIB_EXTRA_PACKAGES="jq yq mdadm lvm2 curl parted util-linux \
+    squashfs-tools xfsprogs dosfstools grub2-efi-aa64 grub2-tools rsync"
+```
+
+## Installing the Hardware Manager
+
+The hardware manager must be placed in the IPA hardware managers directory
+and registered in `setup.cfg`.
+
+**File location:**
+
+```text
+ironic_python_agent/hardware_managers/deb_oci_efi_lvm.py
+```
+
+**setup.cfg entry point:**
+
+Add the following entry to the `ironic_python_agent.hardware_managers`
+section in `setup.cfg`:
+
+```ini
+[entry_points]
+ironic_python_agent.hardware_managers =
+    deb_oci_efi_lvm = ironic_python_agent.hardware_managers.deb_oci_efi_lvm:DebOCIEFILVMHardwareManager
+```
+
+This registers the hardware manager as a plugin, allowing IPA to
+discover and load it at runtime.
+
+<!-- markdownlint-disable MD033 -->
+
+<details>
+  <summary>Complete hardware manager source code (deb_oci_efi_lvm.py)</summary>
+  <div markdown="1">
+
+<!-- markdownlint-enable MD033 -->
+
+```python
```

Comment:
It could be useful to split this very large code block into smaller parts with separate commentary, and potentially organize it in a reversed way: start by showing the highest level code (the deploy step itself), then descend into details. You may even decide to omit helper functions, pointing at your repository instead.